### PR TITLE
Don't special-case "help" in the parser and have user functions return ChannelMessages

### PR DESCRIPTION
**Pull request by ento**
for #8

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,7 +1,8 @@
 language: python
 cache: pip
 python:
-  - "nightly" # currently points to 3.6-dev
+  # see https://docs.travis-ci.com/user/languages/python/ for options
+  - "nightly"
```
Comment (eeue56) [inline .travis.yml]: Right 3.7 is the same as 3.6-dev. Was hoping to tie it to 3.6 beta instead, but since that's not a travis option yet I figured it would be okay to wait til 3.6 final

File: slack_today_i_did/parser.py [hunk 2]
```diff
@@ -143,11 +221,14 @@ def mismatching_args_messages(action, annotations, args) -> str:
 def mismatching_types_messages(action, annotations, args) -> str:
     messages = []
 
-    for ((arg, arg_type), (arg_name, annotation)) in zip(args, annotations.items()):  # noqa: E501
-        if arg_type != annotation:
+    for (arg, (arg_name, annotation)) in zip(args, annotations.items()):
+        if annotation != Any and not isinstance(annotation, TypeVar):
```
Comment (eeue56) [inline slack_today_i_did/parser.py]: When the annotation is something in particular, but isn't one of the type variables, then we want to skip it.

In what cases does this apply?

File: slack_today_i_did/parser.py [hunk 3]
```diff
@@ -59,50 +80,107 @@ def tokenize(text, known_tokens):
...
+        all_errors.extend(argument_errors)
+        return FuncResult(None, None, [], all_errors)
+
+    try:
+        return FuncResult(action(*evaluated_args), action, evaluated_args, all_errors)
+    except Exception as e:
+        return FuncResult(None, None, [], [exception_error_messages([(func_call.func_name, e)])])
+
+
+def evaluate_args(known_functions, args, default_args=[]):
```
Comment (eeue56) [inline slack_today_i_did/parser.py]: :thought_balloon: maybe worth adding type annotations here to explain what these things should be

File: slack_today_i_did/parser.py [hunk 4]
```diff
@@ -1,3 +1,24 @@
...
+    'FuncResult',
+    [('result', Any), ('action', Callable), ('args', List[Any]), ('errors', List[str])])
+
+
+def metafunc(fn):
+    fn.is_metafunc = True
+    return fn
+
+
+def is_metafunc(fn):
```
Comment (eeue56) [inline slack_today_i_did/parser.py]: :thought_balloon: can this not be moved out into types instead of adding attrs?

**Comment by eeue56** [comment]
This is awesome! I have a couple of questions about the implementation that I couldn't figure out just from reading, but otherwise this is perfect to merge :100:

File: .travis.yml (hunk 1 above)
Comment (ento) [inline .travis.yml]: How's 3.7 the same as 3.6-dev? (I don't know much about Python's dev cycle and after reading https://docs.python.org/devguide/devcycle.html, it seems to be saying that the default branch == 3.7 != 3.6)

Is 3.6-dev https://hg.python.org/cpython/shortlog/3.6?

File: slack_today_i_did/parser.py (hunk 2 above)
Comment (ento) [inline slack_today_i_did/parser.py]: For all cases we want to actually type check :p and this is actually a bug. Thanks for spotting it. Fixed in 84551f7

File: slack_today_i_did/parser.py (hunk 3 above)
Comment (ento) [inline slack_today_i_did/parser.py]: +1 added annotations be7634e

File: slack_today_i_did/parser.py (hunk 4 above)
Comment (ento) [inline slack_today_i_did/parser.py]: do you mean wrapping `fn` in a special type, or looking at argument annotations? either way, going to make a separate PR

**Comment by ento** [comment]
pushed a non-trivial amount of changes but merging to prevent drift

File: .travis.yml (hunk 1 above)
Comment (eeue56) [inline .travis.yml]: I'm not sure on what travis-ci considers nightly, but since no tags nor branches exist for 3.7 yet (nor will they for a little bit), I can only assume that they mean `tip` (which is https://hg.python.org/cpython/shortlog/default ) 

Right now, tip is the same as 3.6 since 3.6 is still in beta mode and 3.7 hasn't had any PEPs planned against it yet

File: slack_today_i_did/parser.py (hunk 4 above)
Comment (eeue56) [inline slack_today_i_did/parser.py]: Right - I meant making it so that the parser can basically pattern match on the type of function. E.g 

```
case func of 
  MetaFunc fn -> ..
  Func fn -> ..
  Constant const -> ..
```
